### PR TITLE
Viewmodel weapon XYZ location changer

### PR DESCRIPTION
**Pull request by RyDeem**
Decided to revive my old viewmodel mover project (based on #584), thanks #2019 for the inspiration.

menu1
menu2

Changes the location of the viewmodel onscreen in a kind of neat menu.

Code may be long or something, but this gets the job done quite good.

Warning:
You will lose your current viewmodel settings, so save them or write them down if you want to not forget what they were.

Notes:
1. Dual berretas/Elites are locked to 0,0,0, currently as they look wierd if moved, might change this later
2. C4/Bomb is locked to zeroes as well.
3. when unloading cheat in misc tab "unhook" button viewmodel gets reset to 0, and so does cl_righthand to 1.
(here we would need a system to get the viewmodel values before inject (or enabling viewmodel mover) and set them back to defaults when needed.
As currently this only sets the XYZ location to 0.0.0 when disabling this feature in the GUI and unhooking the cheat, not what it was before/user settings.)
4.  Added grenades and misc/dangerzone equipables to WeaponId.h to make my life more easy.
5. This is mainly aimed at people who use cl_righthand 1 all the time, i say this because it resets to 1 when disabling the function/unhooking, this needs to change to the users set values.

Can someone help me optimize this? it works in local server if inject after connected but crashes sometimes otherwise in casual any maybe comp too

### CODE REVIEW

**Comment by KernCS** [comment]
Nice! maybe see if you can add [roll?](https://www.unknowncheats.me/forum/counterstrike-global-offensive/325940-viewmodel-roll.html)

**Comment by RyDeem** [comment]
> Nice! maybe see if you can add [roll?](https://www.unknowncheats.me/forum/counterstrike-global-offensive/325940-viewmodel-roll.html)

I got this far but i cant figure out how to full port it. (this is not functional)

Can someone take a look?

```
        if (interfaces->engine->isInGame() && localPlayer && localPlayer->isWeapon())
        {
            auto ViewModel = reinterpret_cast<Entity*>(interfaces->entityList->getEntityFromHandle(localPlayer->viewModel()));
            if (ViewModel)
            {
                auto vsView = Utils::GetVFunc<13, CViewSetup*>(ecx);
                auto eyeAng = vsView->angles;
                eyeAng.z -= 90.f;

                ViewModel->SetAbsAngles(eyeAng);
            }
        }
```

**Comment by NekoRem** [comment]
pretty sure you can reduce the amount of code by using the config system similar to the chams. idk ive not looked at the config system recently and i know its changed slightly

**Comment by NekoRem** [comment]
also i'd move this to somewhere else so it doesn't get looped constantly
```
static ConVar* viewmodel_x = interfaces->cvar->findVar("viewmodel_offset_x");
    static ConVar* viewmodel_y = interfaces->cvar->findVar("viewmodel_offset_y");
    static ConVar* viewmodel_z = interfaces->cvar->findVar("viewmodel_offset_z");
    static ConVar* cl_righthand = interfaces->cvar->findVar("cl_righthand");
    static ConVar* sv_minspec = interfaces->cvar->findVar("sv_competitive_minspec");
```

**Comment by RyDeem** [comment]
Thank you for looking in to this Rem.

config rework would be needed yes, even more if i want to add more types of guns, or each gun independent like aimbot does.

where would i move the static cons then?

**Comment by braken1** [comment]
Hello, i use other osiris version which has other cfg saving method i like this version because it has jb and better esp than new osiris
i tried to redo the config from your post but still have errors, any help pls?


 if (visualsJson.isMember("Playermodel CT")); visuals.playerModelCT = visualsJson["Playermodel CT"].asInt();
        if (visualsJson.isMember("ViewmodelXYZ")); v.viewmodelxyz = visualsJson["ViewmodelXYZ"].asBool();
        if (visualsJson.isMember("ViewmodelXYZ cl_right")); v.viewmodel_clright = visualsJson["ViewmodelXYZ cl_right"].asBool();
        if (visualsJson.isMember("ViewmodelXYZ X")); v.viewmodel_x = visualsJson["ViewmodelXYZ X"].asBool();
        if (visualsJson.isMember("ViewmodelXYZ Y")); v.viewmodel_y = visualsJson["ViewmodelXYZ Y"].asBool();
        if (visualsJson.isMember("ViewmodelXYZ Z")); v.viewmodel_z = visualsJson["ViewmodelXYZ Z"].asBool();
        if (visualsJson.isMember("Color correction")) {
            const auto& cc = visualsJson["Color correction"];


visualsJson["Playermodel T"] = visuals.playerModelT;
        visualsJson["Playermodel CT"] = visuals.playerModelCT;
        visualsJson["ViewmodelXYZ"] = v.viewmodelxyz;
        visualsJson["ViewmodelXYZ cl_right"] = v.viewmodel_clright;
        visualsJson["ViewmodelXYZ X"] = v.viewmodel_x;
        visualsJson["ViewmodelXYZ Y"] = v.viewmodel_y;
        visualsJson["ViewmodelXYZ Z"] = v.viewmodel_z;

**Comment by braken1** [comment]
> Hello, i use other osiris version which has other cfg saving method i like this version because it has jb and better esp than new osiris
> i tried to redo the config from your post but still have errors, any help pls?
> 
> if (visualsJson.isMember("Playermodel CT")); visuals.playerModelCT = visualsJson["Playermodel CT"].asInt();
> if (visualsJson.isMember("ViewmodelXYZ")); v.viewmodelxyz = visualsJson["ViewmodelXYZ"].asBool();
> if (visualsJson.isMember("ViewmodelXYZ cl_right")); v.viewmodel_clright = visualsJson["ViewmodelXYZ cl_right"].asBool();
> if (visualsJson.isMember("ViewmodelXYZ X")); v.viewmodel_x = visualsJson["ViewmodelXYZ X"].asBool();
> if (visualsJson.isMember("ViewmodelXYZ Y")); v.viewmodel_y = visualsJson["ViewmodelXYZ Y"].asBool();
> if (visualsJson.isMember("ViewmodelXYZ Z")); v.viewmodel_z = visualsJson["ViewmodelXYZ Z"].asBool();
> if (visualsJson.isMember("Color correction")) {
> const auto& cc = visualsJson["Color correction"];
> 
> visualsJson["Playermodel T"] = visuals.playerModelT;
> visualsJson["Playermodel CT"] = visuals.playerModelCT;
> visualsJson["ViewmodelXYZ"] = v.viewmodelxyz;
> visualsJson["ViewmodelXYZ cl_right"] = v.viewmodel_clright;
> visualsJson["ViewmodelXYZ X"] = v.viewmodel_x;
> visualsJson["ViewmodelXYZ Y"] = v.viewmodel_y;
> visualsJson["ViewmodelXYZ Z"] = v.viewmodel_z;

oh nvm i fixed it by replading v.viewmodel_z by visuals.viewmodel_z

**Comment by braken1** [comment]
haha i pasted the first version which dont have the opposite cl_righthand feature on knife
ggggg

**Comment by braken1** [comment]
would you mind creating a simple opposite hand knife feature -_- im dumb and cant make this code work, the first one worked but didnt saved to cfg

**Comment by RyDeem** [comment]
> would you mind creating a simple opposite hand knife feature -_- im dumb and cant make this code work, the first one worked but didnt saved to cfg

```
///////// in Visuals.cpp for example
void Visuals::KnifeSwapper(FrameStage stage) noexcept{

static ConVar* cl_righthand = interfaces->cvar->findVar("cl_righthand");
const bool config_righthand = config->visuals.viewmodelXYZ.viewmodel_clright;

auto activeWeapon = localPlayer->getActiveWeapon();

if (!localPlayer->isAlive())
{
    if (!localPlayer->getObserverTarget())
        return;
    if (localPlayer->getObserverMode() != ObsMode::InEye)
        return;
    const auto activeWeapon = localPlayer->getObserverTarget()->getActiveWeapon();
}

const auto classid = activeWeapon->getClientClass()->classId;
if (classid == ClassId::Knife) { 
    if(cl_righthand->getInt() != config_righthand ){
        cl_righthand = config_righthand ;
    }
}
}

/////// in Gui.cpp

ImGui::Checkbox("cl_righthand", &config->visuals.viewmodelXYZ.viewmodel_clright);
```

i hope that helps, figure out the hooks and .h file yourself if you can

**Comment by braken1** [comment]
> > would you mind creating a simple opposite hand knife feature -_- im dumb and cant make this code work, the first one worked but didnt saved to cfg
> 
> ```
> ///////// in Visuals.cpp for example
> void Visuals::KnifeSwapper() noexcept{
> 
> static ConVar* cl_righthand = interfaces->cvar->findVar("cl_righthand");
> const bool config_righthand = config->visuals.viewmodelXYZ.viewmodel_clright;
> 
> auto activeWeapon = localPlayer->getActiveWeapon();
> 
> if (!localPlayer->isAlive())
> {
>     if (!localPlayer->getObserverTarget())
>         return;
>     if (localPlayer->getObserverMode() != ObsMode::InEye)
>         return;
>     const auto activeWeapon = localPlayer->getObserverTarget()->getActiveWeapon();
> }
> 
> const auto classid = activeWeapon->getClientClass()->classId;
> if (classid == ClassId::Knife) { 
>     if(cl_righthand->getInt() != config_righthand ){
>         cl_righthand = config_righthand ;
>     }
> }
> }
> 
> /////// in Gui.cpp
> 
> ImGui::Checkbox("Knife Swap", &config->visuals.viewmodelXYZ.viewmodel_clright);
> ```
> 
> i hope that helps, figure out the hooks and .h file yourself if you can

ill try it thanks

**Comment by RyDeem** [comment]
I hope it works for you i did not test it wrote it just now

**Comment by braken1** [comment]
> I hope it works for you i did not test it wrote it just now, so you might have to fix some stuff

i guess this only works with xyz changes const bool config_righthand = config->visuals.viewmodelXYZ.viewmodel_clright; 
not as code itself?

**Comment by RyDeem** [comment]
> > I hope it works for you i did not test it wrote it just now, so you might have to fix some stuff
> 
> i guess this only works with xyz changer const bool config_righthand = config->visuals.viewmodelXYZ.viewmodel_clright;
> not as code itself?

Yes but make that config path yourself, i just made it to be close to the end product.
Like

const bool config_righthand = config->visuals.Knife_cl_clright;

the config is not hard currently, in config.h in 
```
    struct Visuals {
bool config_righthand { false };
}
```

and in config.cpp 

static void from_json(const json& j, Config::Visuals& v) //This is to LOAD config.
{

    read<value_t::boolean>(j, "Config Righthand", v.config_righthand ); (THIS IS FOR bool )

    read_number(j, "Config Righthand", v.config_righthand ); (THIS IS FOR int , float , etc )
}

static void to_json(json& j, const Config::Misc& o) //This is to save config.
{
WRITE("Config Righthand", config_righthand);
}
-----------------------------
so figur out wich one you need (bool)

and this code is not completekly copy pastable, so just try to figure it out, coding is like a puzzle game when you dont know the basics imho, so beat the puzzle my friend. :)

File: Osiris/GUI.cpp [hunk 1]
```diff
@@ -924,6 +924,82 @@ void GUI::renderVisualsWindow(bool contentOnly) noexcept
     ImGui::Checkbox("No grass", &config->visuals.noGrass);
     ImGui::Checkbox("No shadows", &config->visuals.noShadows);
     ImGui::Checkbox("Wireframe smoke", &config->visuals.wireframeSmoke);
+    ImGui::Checkbox("Viewmodel XYZ", &config->visuals.viewmodelXYZ.enabled);
+    ImGui::SameLine();
+    bool viewmodelPopup = ImGui::Button("Edit");
```
Comment (playday3008) [inline Osiris/GUI.cpp:929]: mb better
```cpp
if (ImGui::Button("Edit")
    ImGui::OpenPopup("##viewmodelPopup");
```

Comment (RyDeem) [reply Osiris/GUI.cpp]: pushed an update addressing this issue, should look better now.

**Comment by deboogergithub** [comment]
https://cdn.discordapp.com/attachments/724638912248807465/740312005520523364/Screenshot_2020-08-04-22-55-27-834_com.android.chrome.jpg

im done

**Comment by danielkrupinski** [comment]
`getWeaponIndex()` and `getWeaponClass()` are used in aimbot and triggerbot, it crashes because of modifications you've made to those functions.

**Comment by RyDeem** [comment]
Sadly it still crashes even without the changes to getWeaponIndex() and getWeaponClass().

i did not change the existing funtion definitions, just added new ones, how would that crash? (trying to understand )

**Comment by braken1** [comment]
hey, maybe im missing something but i still get 2 errors 
i tried to fix 1 and had another 200 so i am asking for help 
image

**Comment by RyDeem** [comment]
1st error looks like you need a pointer to the function viewmodelxyz in visuals.h
2nd looks like yuo are missing a { before a }

**Comment by SapphireCodenz** [comment]
image

**Comment by SapphireCodenz** [comment]
crashes when afk/tabbed out

**Pull request closed without merging**